File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-42.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {  
        let count = {}
        let res = 0 
        let curMax = 0

        let l = 0 
        for (let r = 0; r < s.length; r ++) {
            let cur = s[r] 

            count[cur] = ( count[cur] || 0 ) + 1  
            curMax = Math.max(count[cur], curMax)

            while(((r - l + 1) - curMax) > k) {
                let deleted = s[l] 
                count[deleted] -- 
                l ++
            } 


            res = Math.max(res, (r - l + 1))
        } 

        return res
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-42.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        c = set()  
        res = 0

        l = 0 
        for r in range(len(s)):  
            cur = s[r] 
            while cur in c:
                deleted = s[l] 
                c.remove(deleted) 
                l += 1 

            c.add(cur)
            res = max(res, r - l + 1) 

        return res